ledger-tool/src/main.rs

```diff
@@ -4104,16 +4086,24 @@ fn main() {
                         Box::new(stdout())
                     };
 
-                slot_hash
+                iter.take(num_roots)
+                    .filter(|slot| *slot <= max_height as u64)
+                    .map(|slot| {
+                        let blockhash = blockstore
+                            .get_slot_entries(slot, 0)
+                            .unwrap()
+                            .last()
+                            .unwrap()
+                            .hash;
+                        (slot, blockhash)
+                    })
+                    .collect::<Vec<_>>()
```

Comment:
This `.collect()` is necessary; removing it results in following compile errors (with or without `.into_iter()`)
```
error[E0277]: the trait bound `impl Iterator<Item = u64>: DoubleEndedIterator` is not satisfied
    --> ledger-tool/src/main.rs:4089:17
     |
4089 | /                 iter.take(num_roots)
4090 | |                     .filter(|slot| *slot <= max_height as u64)
4091 | |                     .map(|slot| {
4092 | |                         let blockhash = blockstore
...    |
4098 | |                         (slot, blockhash)
4099 | |                     })
     | |______________________^ the trait `DoubleEndedIterator` is not implemented for `impl Iterator<Item = u64>`
4100 |                       .rev()
     |                        --- required by a bound introduced by this call
     |
     = help: the trait `DoubleEndedIterator` is implemented for `std::iter::Map<I, F>`
     = note: required for `std::iter::Take<impl Iterator<Item = u64>>` to implement `DoubleEndedIterator`
     = note: 2 redundant requirements hidden
     = note: required for `std::iter::Map<Filter<std::iter::Take<impl Iterator<Item = u64>>, [closure@ledger-tool/src/main.rs:4090:29: 4090:35]>, [closure@ledger-tool/src/main.rs:4091:26: 4091:32]>` to implement `DoubleEndedIterator`
note: required by a bound in `std::iter::Iterator::rev`

error[E0277]: the trait bound `impl Iterator<Item = u64>: ExactSizeIterator` is not satisfied
    --> ledger-tool/src/main.rs:4089:17
     |
4089 | /                 iter.take(num_roots)
4090 | |                     .filter(|slot| *slot <= max_height as u64)
4091 | |                     .map(|slot| {
4092 | |                         let blockhash = blockstore
...    |
4098 | |                         (slot, blockhash)
4099 | |                     })
     | |______________________^ the trait `ExactSizeIterator` is not implemented for `impl Iterator<Item = u64>`
4100 |                       .rev()
     |                        --- required by a bound introduced by this call
     |
     = help: the following other types implement trait `ExactSizeIterator`:
               &mut I
               Args
               ArgsOs
               ArrayChunksMut<'_, T, N>
               ArrayWindows<'_, T, N>
               Box<I, A>
               ByteRecordIter<'r>
               ChunksExact<'_, T>
             and 274 others
     = note: required for `std::iter::Take<impl Iterator<Item = u64>>` to implement `DoubleEndedIterator`
     = note: 2 redundant requirements hidden
     = note: required for `std::iter::Map<Filter<std::iter::Take<impl Iterator<Item = u64>>, [closure@ledger-tool/src/main.rs:4090:29: 4090:35]>, [closure@ledger-tool/src/main.rs:4091:26: 4091:32]>` to implement `DoubleEndedIterator`
note: required by a bound in `std::iter::Iterator::rev`
```

Reply:
Checked that as well.
Iterator returned by the `open_blockstore` is just an [`Iterator`](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html), not a [`DoubleEndedIterator`](https://doc.rust-lang.org/stable/std/iter/trait.DoubleEndedIterator.html) that [`rev()`](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.rev) needs.
See the `rev()` method bounds:

```rust
    fn rev(self) -> Rev<Self>
    where
        Self: Sized + DoubleEndedIterator,
```

Seems reasonable, as `Iteator` only provides an ability to get items one by one and only once.
So one would need to either store them, as `collect::<Vec<_>>()` would do, or provide reverse iteration in some other way.

Cursory investigation shows, `open_blockstore` binds to the RocksDB API.
And there exists a way to iterate in reverse, via `reversed_rooted_slot_iterator()`, but then one needs to provide an end point.
And that seems to be harder to calculate, given the current tool arguments.
Guess, it is probably good enough as it is :)